File: test/domain/QueryMatcher.scala
package domain

import util.CompositeMatcher
import org.hamcrest.{Description, TypeSafeMatcher}

object QueryMatcher {
    def aQuery = new QueryMatcher()
}

class QueryMatcher extends CompositeMatcher[Query] {
    val typeName = "a Query"

    def withRegEx(regex: String) = {
        add(new TypeSafeMatcher[Query]() {
            def matchesSafely(query: Query): Boolean = query.toRegEx == regex


            def describeTo(description: Description) {
                description.appendText(s"where the regex is $regex")
            }
        })
        this
    }
}

File: test/domain/QueryMatcher.java
package domain;

import org.hamcrest.Description;
import org.hamcrest.TypeSafeMatcher;
import util.CompositeMatcher;

public class QueryMatcher extends CompositeMatcher<Query> {

    private QueryMatcher(){
        super();
    }

    public static QueryMatcher aQuery() {
        return new QueryMatcher();
    }

    public QueryMatcher withRegEx(final String regEx) {
        add(new TypeSafeMatcher<Query>() {
            @Override
            public boolean matchesSafely(Query query){
                return query.queryString().value.equals(regEx);
            }

            @Override
            public void describeTo(Description description) {
                description.appendText("Where the target is " + regEx);
            }
        });
        return this;
    }
}
